### PR TITLE
[Feature] Hooks and Buffers for TensorDictParams

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Introduces 2 features for TensorDictParams:
- `pre_get_hooks` allow users to register operations to be executed every time a `get` operation is executed (incl. values, items, `__getitem__`, `get`, `_get_str`, `_get_tuple`). For instance, this allows to create an instance of TensorDictParams that gets the `.data` attribute of another one while keeping the content similar after casting to device or dtype.
- Introduces a `Buffer` class, similar to `nn.Parameter` but without gradient by default. If a tensor is not a parameter in TensorDictParam, it will be a `Buffer`. This change allows us to call `TensorDictParams.to(...)` while being sure that every leaf of the structure has the same identity before and after the operation, as well as the buffers in `_buffers`. Not using this class would result in the content of `_buffers` being updated but not the content of the tensordict, thereby breaking the sync between the two.

cc @matteobettini

### CODE REVIEW

File: tensordict/nn/params.py [hunk 1]
```diff
@@ -196,15 +264,45 @@ class TensorDictParams(TensorDictBase, nn.Module):
...
+    def _reset_params(self):
+        parameters = self._param_td
+        param_keys = [
+            key
+            for key, value in parameters.items(True, True)
+            if isinstance(value, nn.Parameter)
+        ]
+        buffer_keys = [
+            key
+            for key, value in parameters.items(True, True)
```
Comment (matteobettini) [inline tensordict/nn/params.py:297]: here we are traversing the whole params tree twice.
if this function is called in other parts other than init it might be worth traversing just once

File: tensordict/nn/params.py [hunk 2]
```diff
@@ -69,19 +108,38 @@ def new_func(_self, *args, **kwargs):
...
             return out
 
         return new_func
 
 
+def _get_post_hook(func):
+    @wraps(func)
+    def new_func(self, *args, **kwargs):
+        out = func(self, *args, **kwargs)
+        if not _is_tensor_collection(type(out)):
```
Comment (matteobettini) [inline tensordict/nn/params.py:133]: so hooks are only for leaves? why?

File: tensordict/nn/params.py [hunk 3]
```diff
@@ -567,6 +684,120 @@ def create_nested(self, key):
...
+        self, include_nested: bool = False, leaves_only: bool = False
+    ) -> Iterator[CompatibleType]:
+        for k, v in self._param_td.items(include_nested, leaves_only):
+            if _is_tensor_collection(type(v)):
+                yield k, v
+                continue
+            for hook in self._get_post_hook:
+                new_v = hook(self, v)
+                if new_v is not None:
+                    v = new_v
```
Comment (matteobettini) [inline tensordict/nn/params.py:710]: this code is repeated 3 times. maybe consider a function `_apply_hooks`

**Review by matteobettini — APPROVED** [review]
LGTM, some nits

File: tensordict/nn/params.py (hunk 2 above)
Comment (vmoens) [reply tensordict/nn/params.py:133]: it's applied recursively
If I do something like `hook = lambda x: x*0` that won't work on tensordict instances
